baqend/_getFMPData.ts

```diff
@@ -0,0 +1,209 @@
+    candidate.wptFMP = null
+    return candidate
+  }
+
+  let lastVisualProgress = data[0][1]
+  for (let i = 1; i < data.length; i += 1) {
+    const [time, visualProgress] = data[i]
+    const diff = visualProgress - lastVisualProgress
+    lastVisualProgress = visualProgress
+
+    const candidate = new db.Candidate()
+    candidate.visualCompleteness = visualProgress
+    candidate.deltaVC = diff
+    candidate.startTime = time * 1000 - 100
+    candidate.endTime = time * 1000
+    candidate.wptFMP = null
+    diffs.push(candidate)
+  }
+
+  return diffs
+}
+
+/**
+ * Calculates ΔVisualCompleteness from the WebPagetest
+ */
+async function prepareCandidates(db: baqend, testId: string, runIndex: string): Promise<model.Candidate[]> {
+  const url = `http://${credentials.wpt_dns}/video/compare.php?tests=${testId}-r:${runIndex}-c:0`
+  const response = await fetch(url)
+  const htmlString = await response.text()
+  const data = getDataFromHtml(htmlString)
+  db.log.info('Found data for FMP calculation', { data })
+
+  return firstDeviation(db, data)
+}
+
+/**
+ * Chooses up to five candidates based on the given deltas.
+ */
+function chooseTopCandidates(db: baqend, candidates: model.Candidate[]): model.Candidate[] {
+  // Find the five highest ΔVCs
+  const topCandidates = candidates
+    .sort(({ deltaVC: a }, { deltaVC: b }) => b - a)
+    .slice(0, 5)
+
+  db.log.info('Candidates for FMP found', { topCandidates })
+  return topCandidates
+}
+
+/**
+ * Gets a candidate that matches the WPT FMP or null.
+ */
+function getValidCandidate(wptFMP: number, candidates: model.Candidate[]): model.Candidate|null {
+  if (wptFMP <= 0) {
+    return null
+  }
+
+  if (candidates.length === 1) {
+    const { startTime, endTime } = candidates[0]
+    const matchesWPT = Math.abs(startTime - wptFMP) <= 100 || Math.abs(endTime - wptFMP) <= 100
+    return matchesWPT ? candidates[0] : null
+  }
+
+  const matchingCandidate = candidates.find(candidate => {
+    return Math.abs(candidate.startTime - wptFMP) <= 100 ||
+      Math.abs(candidate.endTime - wptFMP) <= 100
+  })
+
+  return matchingCandidate || null
+}
+
+/**
+ * Calculates the candidate based on the given deltas.
+ */
+function calculateCandidate(db: baqend, candidates: model.Candidate[], wptFMP: number): model.Candidate {
+  const suggestedCandidate = new db.Candidate()
+  if (candidates.length === 1) {
+    return candidates[0]
+  }
+
+  for(const i in candidates) {
+    const { deltaVC } = candidates[i]
+
+    // stop loop if the visual progress is negative => FMP is last highest diff
+    if (deltaVC < 0) {
+      break
+    }
+
+    if (typeof suggestedCandidate.deltaVC === 'undefined' || deltaVC > suggestedCandidate.deltaVC) {
+      Object.assign(suggestedCandidate, candidates[i].toJSON())
+    }
+
+    if (suggestedCandidate.deltaVC >= 50) {
+      break
+    }
+  }
+
+  if (Math.abs(suggestedCandidate.startTime - wptFMP) <= 100 || Math.abs(suggestedCandidate.endTime - wptFMP) <= 100) {
```

Comment:
Isn't this redundant? If there were a candidate near the WPT FMP, we would have chosen the wpt FMP, right?